lib/src/Pos/Diffusion/Full.hs

```diff
@@ -0,0 +1,449 @@
+import qualified System.Metrics as Monitoring
+import           System.Random (newStdGen)
+import           System.Wlog (WithLogger, CanLog, logError, getLoggerName, usingLoggerName)
+
+import           Pos.Block.Network.Types (MsgGetHeaders, MsgHeaders, MsgGetBlocks, MsgBlock)
+import           Pos.Communication (NodeId, VerInfo (..), PeerData, PackingType, EnqueueMsg, makeEnqueueMsg, bipPacking, Listener, MkListeners (..), HandlerSpecs, InSpecs (..), OutSpecs (..), createOutSpecs, toOutSpecs, convH, InvOrDataTK, MsgSubscribe, makeSendActions, SendActions, Msg)
+import           Pos.Communication.Limits (HasAdoptedBlockVersionData (..))
+import           Pos.Communication.Relay.Logic (invReqDataFlowTK)
+import           Pos.Communication.Util (wrapListener)
+import           Pos.Configuration (HasNodeConfiguration, conversationEstablishTimeout, networkConnectionTimeout)
+import           Pos.Core.Block (Block, MainBlockHeader, BlockHeader)
+import           Pos.Core.Configuration (protocolMagic)
+import           Pos.Core.Types (HeaderHash, BlockVersionData (..), StakeholderId, ProxySKHeavy)
+import           Pos.Core.Txp (TxAux)
+import           Pos.Core.Update (UpId, UpdateVote, UpdateProposal)
+import           Pos.Crypto.Configuration (ProtocolMagic (..))
+import           Pos.DHT.Real (KademliaDHTInstance, KademliaParams (..), startDHTInstance, stopDHTInstance)
+import qualified Pos.Diffusion.Full.Block      as Diffusion.Block
+import qualified Pos.Diffusion.Full.Delegation as Diffusion.Delegation
+import qualified Pos.Diffusion.Full.Ssc        as Diffusion.Ssc
+import qualified Pos.Diffusion.Full.Txp        as Diffusion.Txp
+import qualified Pos.Diffusion.Full.Update     as Diffusion.Update
+import           Pos.Diffusion.Full.Types  (DiffusionWorkMode)
+import           Pos.Diffusion.Types (Diffusion (..), DiffusionLayer (..), GetBlocksError (..))
+import           Pos.Logic.Types (Logic (..))
+import           Pos.Network.CLI (NetworkConfigOpts (..), intNetworkConfigOpts)
+import           Pos.Network.Types (NetworkConfig (..), Topology (..), Bucket, initQueue,
+                                    topologySubscribers, SubscriptionWorker (..),
+                                    topologySubscriptionWorker)
+import           Pos.Ssc.Message (MCOpening, MCShares, MCCommitment, MCVssCertificate)
+import           Pos.Diffusion.Subscription.Common (subscriptionListeners)
+import           Pos.Diffusion.Subscription.Dns (dnsSubscriptionWorker)
+import           Pos.Diffusion.Subscription.Dht (dhtSubscriptionWorker)
+import           Pos.Update.Configuration (lastKnownBlockVersion)
+import           Pos.Util.OutboundQueue (EnqueuedConversation (..))
+
+{-# ANN module ("HLint: ignore Reduce duplication" :: Text) #-}
+
+-- Orphan instance to get the adopted block version data through reflection.
+instance (Given (m BlockVersionData)) => HasAdoptedBlockVersionData m where
```

Comment:
This orphan instance doesn't look good to me, it conflicts with other existing instances. With what `m` instantiations is it used? Would it be possible to specialize `m`? I.e. `instance Given (Production BlockVersionData) => HasAdoptedBlockVersionData Production`

Reply:
The only instance it conflicts with is the `MonadGState`-backed one, used to keep the existing cardano-sl implementation up and running. That one isn't imported here so it's ok (no overlap).

If we specialized it then the full diffusion layer implementation given here would also have to specialize, but we don't want to specialize it. If we did, we'd have to lift / natural transform the monad parameter at use sites and that's cumbersome as you know (negative position `m`).